crates/module-system/module-implementations/sov-test-state-consistency/src/call.rs

```
@@ -0,0 +1,162 @@
+    },
+}
+
+impl<S: Spec> StateConsistency<S> {
+    pub(crate) fn update_value(
+        &mut self,
+        old_check: u64,
+        new: u64,
+        context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        // Get the current value for this sender's address
+        let sender_address = context.sender();
+        let current_value = self.values.get(sender_address, state)?.unwrap_or(0);
+
+        anyhow::ensure!(
+            current_value == old_check,
+            "Current value mismatched: stored value {}, transaction expected {}",
+            current_value,
+            old_check
+        );
+
+        // Update the value for this address
+        self.values.set(sender_address, &new, state)?;
+
+        // Emit event
+        self.emit_event(
+            state,
+            Event::ValueUpdated {
+                address: sender_address.clone(),
+                old_value: current_value,
+                new_value: new,
+            },
+        );
+
+        Ok(())
+    }
+
+    pub(crate) fn update_accessory_state(
+        &mut self,
+        new: u64,
+        _context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        Ok(self.accessory_value.set(&new, state)?)
+    }
+
+    pub(crate) fn assert_block_state(
+        &mut self,
+        expected_visible_slot_number: u64,
+        expected_rollup_height: u64,
+        expected_state_root: Vec<u8>,
+        _context: &Context<S>,
+        state: &mut impl TxState<S>,
+    ) -> Result<()> {
+        let mut mismatches = Vec::<String>::new();
+
+        let rollup_height = state.rollup_height_to_access().get();
+        // Sanity check
+        let saved_rollup_height = self.latest_rollup_height.get(state)?.unwrap_or(0);
+        assert_eq!(rollup_height, saved_rollup_height, "Mismatch between rollup height saved from begin_rollup_block_hook: {saved_rollup_height} and the one read from TxState: {rollup_height}");
```

Here we panic on failure. Is that really the behavior we want? Should this return an error instead?

---

Ah nvm, this is a consistency check in the module not part of the test criteria.  That seems reasonable

---

Yeah, this should never happen: `saved_rollup_height` comes from the begin_rollup_block_hook. The `state` begin hook and the TxState in a transaction should both return the same `rollup_height_to_access()` (unless I'm wrong about this), and if the rollup height changes mid-block it seems like a pretty critical issue.

Since this module is intended for testing, it seems fine to fail ASAP and just crash the rollup if we detect this inconsistency. But even in production this seems like the kind of error that I'd crash the rollup over rather than continue with major state handling bugs.

---

Ah somehow github only showed me your first comment haha. :+1: